flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/CommitterOperatorFactory.java

```diff
@@ -0,0 +1,117 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.operators.sink;
+
+import org.apache.flink.api.connector.sink.Committer;
+import org.apache.flink.api.connector.sink.GlobalCommitter;
+import org.apache.flink.api.connector.sink.Sink;
+import org.apache.flink.core.io.SimpleVersionedSerializer;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperatorFactory;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperatorFactory;
+import org.apache.flink.streaming.api.operators.StreamOperator;
+import org.apache.flink.streaming.api.operators.StreamOperatorParameters;
+
+import java.io.IOException;
+import java.util.Optional;
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * A {@link org.apache.flink.streaming.api.operators.StreamOperatorFactory} for {@link
+ * CommitterOperator}.
+ *
+ * @param <CommT> the type of the committable
+ * @param <GlobalCommT> the type of the global committable
+ */
+public final class CommitterOperatorFactory<CommT, GlobalCommT>
+        extends AbstractStreamOperatorFactory<byte[]>
+        implements OneInputStreamOperatorFactory<byte[], byte[]> {
+
+    private final Sink<?, CommT, ?, GlobalCommT> sink;
+    private final boolean batch;
+
+    public CommitterOperatorFactory(Sink<?, CommT, ?, GlobalCommT> sink, boolean batch) {
+        this.sink = sink;
+        this.batch = batch;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T extends StreamOperator<byte[]>> T createStreamOperator(
+            StreamOperatorParameters<byte[]> parameters) {
+
+        SimpleVersionedSerializer<CommT> committableSerializer =
+                sink.getCommittableSerializer().orElseThrow(this::noSerializerFound);
+        try {
+            CommitterHandler<CommT, GlobalCommT> committerHandler = getGlobalCommitterHandler();
+            if (batch) {
+                Optional<Committer<CommT>> committer = sink.createCommitter();
+                if (committer.isPresent()) {
+                    committerHandler =
+                            new BatchCommitterHandler<>(committer.get(), committerHandler);
+                }
+            }
+
+            checkState(
```

Comment:
Isn't this always failing for batch mode because the batch mode set the committerHandler to `BatchCommitterHandler`?

Reply:
It's always succeeding (`Precondition#checkState` fails on `false`) in your case. However, it could fail if there is no global committer in streaming and it fails if there is neither a committer nor a global committer in batch. But this is a programmatic error in `SinkTranslater`; hence the `checkState`.